plumbing/object/file.go

```diff
@@ -81,7 +81,7 @@ func (iter *FileIter) Next() (*File, error) {
 			return nil, err
 		}
 
-		if entry.Mode.IsDir() {
+		if entry.Mode.IsDir() || entry.Mode == SubmoduleMode {
```

Comment:
Invert the logic here. Just continue if it is not a file, otherwise we are always adding exceptions as they come along.